# Auxiliar Notebook 

- Auxiliary notebook for the transformation of the data that is intended to be used for the project.

## Libraries

In [4]:
import pandas as pd
import numpy as np
import utils as ut

# Rows and columns pandas adjustment
pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', 1000)

## Data extraction

In [7]:
data_raw = pd.read_csv('../DATA/ESGData.csv')
ut.std_nombres_columnas(data_raw)
print(f'Shape: {data_raw.shape}')
data_raw.head(3)

Shape: (16013, 67)


country_name country_code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   

                                      indicator_name     indicator_code  1960  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   NaN   
1            Access to electricity (% of population)     EG.ELC.ACCS.ZS   NaN   
2  Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS   NaN   

   1961  1962  1963  1964  1965  1966  1967  1968  1969      1970      1971  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN       NaN       NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN       NaN       NaN   
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  5.512744  6.667763   

       1972       1973       1974       1975       1976       1977       1978  \
0       NaN        NaN        NaN        NaN        NaN        NaN        NaN   
1       NaN        NaN        NaN        NaN        NaN        NaN        NaN   
2  8.285866  11.445289  23.968493  16.493731  18.342618  18.373684  15.867156   

        1979       1980       1981       1982      1983      1984      1985  \
0        NaN        NaN        NaN        NaN       NaN       NaN       NaN   
1        NaN        NaN        NaN        NaN       NaN       NaN       NaN   
2  32.263267  27.051428  19.370799  10.461897  9.222052  9.104325  7.770835   

       1986      1987    1988      1989      1990      1991      1992  \
0       NaN       NaN     NaN       NaN       NaN       NaN       NaN   
1       NaN       NaN     NaN       NaN       NaN       NaN       NaN   
2  5.104291  6.320428  5.7933  8.195306  8.096032  8.368431  8.451037   

       1993      1994      1995      1996       1997       1998       1999  \
0       NaN       NaN       NaN       NaN        NaN        NaN        NaN   
1       NaN       NaN       NaN       NaN  77.297851  78.302779  78.980861   
2  8.247409  7.712396  8.064746  9.571448   8.379450   5.703252   7.278630   

        2000       2001       2002       2003       2004       2005  \
0  73.702495  74.944860  76.258066  77.120464  78.094440  79.062000   
1  78.479364  80.023047  77.999111  79.306311  79.537184  80.574728   
2  10.316367   8.050232   7.957603   9.539584  10.899546  13.506496   

        2006       2007       2008       2009       2010       2011  \
0  79.851681  80.570536  81.210886  81.809630  82.368101  82.783289   
1  84.621206  82.926878  83.644992  84.270921  86.959991  87.199474   
2  13.688121  12.303601  13.997417   8.721330   9.966140  13.002144   

        2012       2013       2014       2015       2016       2017  \
0  83.120303  83.533457  83.897596  84.171599  84.510171        NaN   
1  87.512260  88.129881  87.275323  88.720097  89.308602  90.283638   
2  12.850205  11.641062  10.437876   6.277652   5.261868   6.385551   

        2018  2019  2020  2050  unnamed:_66  
0        NaN   NaN   NaN   NaN          NaN  
1  89.286856   NaN   NaN   NaN          NaN  
2   8.190277   NaN   NaN   NaN          NaN

In [8]:
# Revision de tipos de datos
ut.univar_analisis(data_raw)[['Variable', 'DataType', 'Total', 'No Missings', 'Missings', '% Missings']]

Variable DataType  Total  No Missings  Missings  % Missings
0     country_name   object  16013        16013         0    0.000000
1     country_code   object  16013        16013         0    0.000000
2   indicator_name   object  16013        16013         0    0.000000
3   indicator_code   object  16013        16013         0    0.000000
4             1960  float64  16013         1352     14661    0.915569
5             1961  float64  16013         1946     14067    0.878474
6             1962  float64  16013         2185     13828    0.863548
7             1963  float64  16013         1959     14054    0.877662
8             1964  float64  16013         1971     14042    0.876913
9             1965  float64  16013         2012     14001    0.874352
10            1966  float64  16013         2013     14000    0.874290
11            1967  float64  16013         2254     13759    0.859239
12            1968  float64  16013         2031     13982    0.873166
13            1969  float64  16013         2043     13970    0.872416
14            1970  float64  16013         3088     12925    0.807157
15            1971  float64  16013         3596     12417    0.775432
16            1972  float64  16013         3839     12174    0.760257
17            1973  float64  16013         3593     12420    0.775620
18            1974  float64  16013         3604     12409    0.774933
19            1975  float64  16013         3942     12071    0.753825
20            1976  float64  16013         3901     12112    0.756385
21            1977  float64  16013         4156     11857    0.740461
22            1978  float64  16013         3913     12100    0.755636
23            1979  float64  16013         3924     12089    0.754949
24            1980  float64  16013         4206     11807    0.737338
25            1981  float64  16013         4235     11778    0.735527
26            1982  float64  16013         4435     11578    0.723038
27            1983  float64  16013         4170     11843    0.739587
28            1984  float64  16013         4209     11804    0.737151
29            1985  float64  16013         4307     11706    0.731031
30            1986  float64  16013         4300     11713    0.731468
31            1987  float64  16013         4604     11409    0.712484
32            1988  float64  16013         4292     11721    0.731968
33            1989  float64  16013         4342     11671    0.728845
34            1990  float64  16013         6596      9417    0.588085
35            1991  float64  16013         6511      9502    0.593393
36            1992  float64  16013         6918      9095    0.567976
37            1993  float64  16013         6697      9316    0.581777
38            1994  float64  16013         6749      9264    0.578530
39            1995  float64  16013         7027      8986    0.561169
40            1996  float64  16013         7963      8050    0.502717
41            1997  float64  16013         7299      8714    0.544183
42            1998  float64  16013         8246      7767    0.485043
43            1999  float64  16013         7220      8793    0.549116
44            2000  float64  16013         9812      6201    0.387248
45            2001  float64  16013         8350      7663    0.478549
46            2002  float64  16013         9909      6104    0.381190
47            2003  float64  16013         9557      6456    0.403172
48            2004  float64  16013         9737      6276    0.391932
49            2005  float64  16013        10067      5946    0.371323
50            2006  float64  16013         9822      6191    0.386623
51            2007  float64  16013        10189      5824    0.363704
52            2008  float64  16013         9777      6236    0.389434
53            2009  float64  16013         9953      6060    0.378443
54            2010  float64  16013        10309      5704    0.356211
55            2011  float64  16013        10066      5947    0.371386
56            2012  

## Data transformation

In [9]:
# Revision de columnas
data_raw.columns

Index(['country_name', 'country_code', 'indicator_name', 'indicator_code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2050',
       'unnamed:_66'],
      dtype='object')

In [12]:
# Transformacion de dataframe
# Eliminando columnas innecesarias
cols_to_drop = ['unnamed:_66']
data_clean = data_raw.drop(columns=cols_to_drop, errors='ignore')

# Reestruturar dataframe
data_transformed = data_clean.melt(
    id_vars=['country_name', 'country_code', 'indicator_name', 'indicator_code'],
    var_name='year',
    value_name='value'
)

# Revision
print('-'*200)
print(f'Shape: {data_transformed.shape}')
print('-'*200)
print('Head')
display(data_transformed.head(5))
print('-'*200)
print('Tail')
display(data_transformed.tail(5))
print('-'*200)

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Shape: (992806, 6)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Head


country_name country_code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      indicator_name     indicator_code  year  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS  1960   
1            Access to electricity (% of population)     EG.ELC.ACCS.ZS  1960   
2  Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS  1960   
3  Adjusted savings: net forest depletion (% of GNI)  NY.ADJ.DFOR.GN.ZS  1960   
4                 Agricultural land (% of land area)     AG.LND.AGRI.ZS  1960   

   value  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Tail


country_name country_code  \
992801     Zimbabwe          ZWE   
992802     Zimbabwe          ZWE   
992803     Zimbabwe          ZWE   
992804     Zimbabwe          ZWE   
992805     Zimbabwe          ZWE   

                                           indicator_name  indicator_code  \
992801  Strength of legal rights index (0=weak to 12=s...  IC.LGL.CRED.XQ   
992802  Terrestrial and marine protected areas (% of t...  ER.PTD.TOTL.ZS   
992803  Unemployment, total (% of total labor force) (...  SL.UEM.TOTL.ZS   
992804  Unmet need for contraception (% of married wom...     SP.UWT.TFRT   
992805                 Voice and Accountability: Estimate          VA.EST   

        year  value  
992801  2050    NaN  
992802  2050    NaN  
992803  2050    NaN  
992804  2050    NaN  
992805  2050    NaN

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


In [13]:
# Revision de numero de paises
data_transformed['country_name'].nunique()

239

In [14]:
# Revision de numero de indicadores
data_transformed['indicator_name'].nunique()

67

In [15]:
# Revision de numero de anios
data_transformed['year'].nunique()

62

In [17]:
# ¿239 paises?
data_transformed['country_name'].unique()

array(['Arab World', 'Caribbean small states',
       'Central Europe and the Baltics', 'Early-demographic dividend',
       'East Asia & Pacific',
       'East Asia & Pacific (excluding high income)',
       'East Asia & Pacific (IDA & IBRD)', 'Euro area',
       'Europe & Central Asia',
       'Europe & Central Asia (excluding high income)',
       'Europe & Central Asia (IDA & IBRD)', 'European Union',
       'Fragile and conflict affected situations',
       'Heavily indebted poor countries (HIPC)', 'High income',
       'IBRD only', 'IDA & IBRD total', 'IDA blend', 'IDA only',
       'IDA total', 'Late-demographic dividend',
       'Latin America & Caribbean',
       'Latin America & Caribbean (excluding high income)',
       'Latin America & Caribbean (IDA & IBRD)',
       'Least developed countries: UN classification',
       'Low & middle income', 'Low income', 'Lower middle income',
       'Middle East & North Africa',
       'Middle East & North Africa (excluding high income)

In [18]:
# Quitando zonas geograficas
non_countries = ['Arab World', 'Caribbean small states',
       'Central Europe and the Baltics', 'Early-demographic dividend',
       'East Asia & Pacific',
       'East Asia & Pacific (excluding high income)',
       'East Asia & Pacific (IDA & IBRD)', 'Euro area',
       'Europe & Central Asia',
       'Europe & Central Asia (excluding high income)',
       'Europe & Central Asia (IDA & IBRD)', 'European Union',
       'Fragile and conflict affected situations',
       'Heavily indebted poor countries (HIPC)', 'High income',
       'IBRD only', 'IDA & IBRD total', 'IDA blend', 'IDA only',
       'IDA total', 'Late-demographic dividend',
       'Latin America & Caribbean',
       'Latin America & Caribbean (excluding high income)',
       'Latin America & Caribbean (IDA & IBRD)',
       'Least developed countries: UN classification',
       'Low & middle income', 'Low income', 'Lower middle income',
       'Middle East & North Africa',
       'Middle East & North Africa (excluding high income)',
       'Middle East & North Africa (IDA & IBRD)', 'Middle income',
       'North America', 'OECD members', 'Other small states',
       'Pacific island small states', 'Post-demographic dividend',
       'Pre-demographic dividend', 'Small states', 'South Asia',
       'South Asia (IDA & IBRD)', 'Sub-Saharan Africa',
       'Sub-Saharan Africa (excluding high income)',
       'Sub-Saharan Africa (IDA & IBRD)', 'Upper middle income', 'World',
      ]

# Solo paises
data_countries_only = data_transformed[~data_transformed['country_name'].isin(non_countries)].reset_index(drop=True)

# Vista rápida
print(f"Shape: {data_countries_only.shape}")
print(f"Numero de paises:\n{data_countries_only['country_name'].nunique()}")

Shape: (801722, 6)
Numero de paises:
193


In [19]:
# Revision del dataframe resultante
print('-'*200)
display(data_countries_only.head(5))
print('-'*200)

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


country_name country_code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                      indicator_name     indicator_code  year  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS  1960   
1            Access to electricity (% of population)     EG.ELC.ACCS.ZS  1960   
2  Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS  1960   
3  Adjusted savings: net forest depletion (% of GNI)  NY.ADJ.DFOR.GN.ZS  1960   
4                 Agricultural land (% of land area)     AG.LND.AGRI.ZS  1960   

   value  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


In [21]:
# Revision
print(f'Numero de paises: {data_countries_only["country_name"].nunique()}')
print(f'Numero de indicadores: {data_countries_only["indicator_name"].nunique()}')
print(f'Numero de anios: {data_countries_only["year"].nunique()}')
print(f'Shape: {data_countries_only.shape}')

Numero de paises: 193
Numero de indicadores: 67
Numero de anios: 62
Shape: (801722, 6)


- Hasta aqui, los numeros cuadran al quedarnos con los 193 paises presentes, cada uno con 67 indicadores a lo largo de 62 años

In [23]:
# Revision de columnas
ut.univar_analisis(data_countries_only)[['Variable', 'DataType', 'Total', 'No Missings', 'Missings', '% Missings']]

Variable DataType   Total  No Missings  Missings  % Missings
0    country_name   object  801722       801722         0    0.000000
1    country_code   object  801722       801722         0    0.000000
2  indicator_name   object  801722       801722         0    0.000000
3  indicator_code   object  801722       801722         0    0.000000
4            year   object  801722       801722         0    0.000000
5           value  float64  801722       288675    513047    0.639931

In [22]:
# Guardando el dataframe limpio
data_countries_only.to_csv('../DATA/ESGData_transformed.csv', index=False)